🐛 fix(e2e): improve bash job wait logic and add diagnostics

rishi-jat opened the pull request:
Fixes #3098

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -142,6 +155,15 @@ kflex ctx --overwrite-existing-context its1
 
 kflex ctx
 
+# Check required kubeconfig contexts exist after they are created
+REQUIRED_CONTEXTS=("$HOSTING_CONTEXT" "its1" "wds1")
+for ctx in "${REQUIRED_CONTEXTS[@]}"; do
+    if ! kubectl config get-contexts "$ctx" >/dev/null 2>&1; then
+        echo "ERROR: kubeconfig context '$ctx' does not exist. Exiting." >&2
+        exit 1
+    fi
+done
+
```
We do not need this. It is clear enough from the existing diagnostics what contexts were missing --- error messages called it out.

MikeSpreitzer commented on test/e2e/bash/use-kubestellar.sh:
```diff
@@ -16,6 +16,10 @@
 set -x # echo so that users can understand what is happening
 set -e # exit on error
 
+# Source the setup-shell.sh to get wait-for-cmd function
+SRC_DIR="$(cd $(dirname "${BASH_SOURCE[0]}") && pwd)"
+source "$SRC_DIR/../common/setup-shell.sh"
+
```
I do not see any use being made of this.

MikeSpreitzer replied:
I still see no reason for adding these lines here.

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -128,7 +128,20 @@ popd
+    kubectl get pods -n its1-system -l job-name=its-with-clusteradm -o wide || true
+    for pod in $(kubectl get pods -n its1-system -l job-name=its-with-clusteradm -o jsonpath='{.items[*].metadata.name}'); do
+        echo "--- Logs for pod $pod ---"
+        kubectl logs -n its1-system "$pod" || true
```
The existing E2E workflow already dumps the logs of failed Pods.
The logs of successful Pods are unlikely to be interesting.

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -128,7 +128,20 @@ popd
+# Robust wait for job.batch/its-with-clusteradm
+if ! kubectl wait -n its1-system job.batch/its-with-clusteradm --for condition=Complete --timeout 400s; then
+    echo "ERROR: Timed out waiting for job.batch/its-with-clusteradm to complete. Dumping job and pod status for debugging..." >&2
+    kubectl get job -n its1-system its-with-clusteradm -o yaml || true
```
Why not dump Jobs in the E2E test workflow, like all the other diagnostic dumping?

rishi-jat replied:
You’re absolutely right! I’ve updated the script to just emit a clear error message on timeout, and left the diagnostic dumping to the E2E workflow for consistency. Thanks for the suggestion 🙌

MikeSpreitzer commented:
/ok-to-test

rishi-jat commented:
@MikeSpreitzer Perfect suggestion! I've added job dumping to the E2E workflow as you recommended. Now both ginkgo-test and test-bash jobs will dump job information consistently, just like they do for pods and deployments.
Much cleaner than having diagnostics scattered across individual scripts. Thanks for the guidance!

rishi-jat commented:
@MikeSpreitzer All requested changes are done removed extra comments and waits, and let kubectl wait fail naturally. Let me know if anything else is needed!

MikeSpreitzer commented on test/e2e/bash/use-kubestellar.sh:
```diff
@@ -16,6 +16,10 @@
 
+# Source the setup-shell.sh to get wait-for-cmd function
+SRC_DIR="$(cd $(dirname "${BASH_SOURCE[0]}") && pwd)"
+source "$SRC_DIR/../common/setup-shell.sh"
```
I see no reason for this command. Nothing in this file makes use of anything defined in the `source`d file. This looks like unnecessary complexity to me.

rishi-jat replied:
Done.

MikeSpreitzer commented:
The title of this PR needs adjusting, right?

rishi-jat commented:
@MikeSpreitzer We can’t remove the source "$SRC_DIR/../common/setup-shell.sh" line because wait-for-cmd is used in this script and is defined in that file. Without sourcing it, the script will fail with “command not found.”

MikeSpreitzer commented:
Let me be a little more precise. Nothing in this PR _adds_ a usage of anything defined in `setup-shell.sh"
`.

There _are_ already uses of `wait-for-cmd` in `setup-kubestellar.sh`. And those uses are _already_ preceded by a `source` of `setup-shell.sh`. You do not need to _add_ **another** `source` of `setup-shell.sh`.

Your test failure is happening for a different reason.

MikeSpreitzer commented:
> I've fixed the flaky E2E bash job issue by improving the job wait logic and making the diagnostics more robust.

This PR no longer makes any change to the job wait logic.

Issue #3098 is NOT about the state-dumping tasks in the workflows, the flake in that Issue is in the test itself.

rishi-jat commented:
@MikeSpreitzer
I've cleaned up the PR to focus only on the core issue. The fix now addresses exactly what you described in #3098

rishi-jat commented:
The fix handles the mixed pod status scenario where the job should be considered successful if at least one pod completed successfully. No workflow changes, no extra improvements - just the essential fix for the flaky E2E test.

rishi-jat commented:
@MikeSpreitzer Ready for review!

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -128,7 +128,18 @@ popd
+kubectl wait -n its1-system job.batch/its-with-clusteradm --for condition=Complete --timeout 400s || {
+    # Check if the job is actually in a successful state
+    job_status=$(kubectl -n its1-system get job its-with-clusteradm -o jsonpath='{.status.conditions[?(@.type=="Complete")].status}' 2>/dev/null || echo "Unknown")
+    if [ "$job_status" = "True" ]; then
+        echo "Job is actually complete, continuing..."
```
Why add another check of the same condition that `kubectl wait` was waiting for?

rishi-jat commented:
@MikeSpreitzer I have a question about handling the flake in #3098.

The issue is that jobs with `backoffLimit: 1` can have one pod fail and another succeed, but `kubectl wait` times out even though the job is actually successful.

I understand you prefer simple code without custom error handling. What's your preferred approach for handling this specific flake?

Options I see:
1. Let `kubectl wait` fail naturally
2. Fix the root cause in the PostCreateHook (reduce backoffLimit to 0?)
3. Something else you'd prefer?

Thanks for your guidance!

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -129,12 +129,36 @@ popd
     kubectl wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-with-clusteradm}=true' --timeout 400s
-    kubectl wait -n its1-system job.batch/its-with-clusteradm --for condition=Complete --timeout 400s
+    kubectl wait -n its1-system job.batch/its-with-clusteradm --for condition=Complete --timeout 400s || {
+        # Handle backoffLimit retry case: check if any pod succeeded
+        if kubectl -n its1-system get pods -l job-name=its-with-clusteradm --field-selector=status.phase=Succeeded -o name | grep -q .; then
+            echo "Job succeeded (some pods completed), continuing..."
+        else
+            echo "Job failed (no pods succeeded)"
+            exit 1
+        fi
+    }
```
I do not understand the point of this. The Job only specifies one Pod. The Job will succeed if any try of that Pod succeeds. This search for a successful Pod is redundant.

MikeSpreitzer commented:
@rishi-jat: what do you mean by the following remark?

> The PostCreateHook templates aren't resolving the KUBECONFIG path correctly.

That's what was fixed in #3163 . In `main` now, the definitions of the environment variable KUBECONFIG get the correctly expanded value.

See https://github.com/kubestellar/kubestellar/blob/ca1f046e90c631a674d79a5483cff67b8d7c830a/docs/content/contribution-guidelines/contributing-inc.md?plain=1#L88

Commit 004c8b0d90244db0a7f0655504af0f106653a19c, currently in the branch of this PR, is a merge from `main`.

MikeSpreitzer commented:
For example, I am running `test/e2e/run-test.sh --test-type bash` . Here is an excerpt from looking at one of the Pods.

```console
mspreitz@mjs13 libvcp % kubectl --context kind-kubeflex get pod -n its1-system install-status-addon-5mj4z -o yaml
...
    env:
    - name: HELM_CONFIG_HOME
      value: /tmp
    - name: HELM_CACHE_HOME
      value: /tmp
    - name: KUBECONFIG
      value: /etc/kube/config-incluster
...
    volumeMounts:
    - mountPath: /etc/kube
      name: kubeconfig
      readOnly: true
...
  volumes:
  - name: kubeconfig
    secret:
      defaultMode: 420
      secretName: vc-vcluster
```

rishi-jat commented:
Got it  thanks for pointing that out. The KUBECONFIG path issue in the PostCreateHook templates is already fixed in main (#3163), so my earlier observation no longer applies. My PR does not rely on that fix, and I’ve updated the branch to reflect that.

rishi-jat commented:
Hi @MikeSpreitzer, I’ve updated the common E2E setup to make early waits context-absolute and added a robust Job wait helper with diagnostics. This should address #3098 and #2201. Please review when you get a chance.

MikeSpreitzer commented on test/e2e/common/setup-shell.sh:
```diff
@@ -48,3 +48,87 @@ function expect-cmd-output() {
+# Usage: wait-for-job-complete <context> <namespace> <job-name> [timeout-seconds]
+# - Considers the Job complete if either:
+#   * .status.conditions has {type: Complete, status: True}, or
+#   * .status.succeeded >= .spec.completions (defaults to 1 when not set), or
```
Why is this necessary?

rishi-jat replied:
You’re right, it’s not strictly necessary  I’ll simplify and rely on kubectl wait as suggested.

MikeSpreitzer commented on test/e2e/common/setup-shell.sh:
```diff
@@ -48,3 +48,87 @@ function expect-cmd-output() {
+# - Considers the Job complete if either:
+#   * .status.conditions has {type: Complete, status: True}, or
+#   * .status.succeeded >= .spec.completions (defaults to 1 when not set), or
+#   * any Pod for the Job has phase Succeeded (helps bridge condition propagation delays).
```
If we are going to start working around propagation delays everywhere, things will become far more complex than necessary. Let's just wait a little longer instead.

MikeSpreitzer commented on test/e2e/common/setup-shell.sh:
```diff
@@ -48,3 +48,87 @@ function expect-cmd-output() {
+        if (( iter >= max_iters )); then
+            echo "Timed out waiting for Job $namespace/$job_name to complete in context $context after ${timeout_sec}s" >&2
+            echo "--- job.describe ---" >&2
+            kubectl --context "$context" -n "$namespace" describe job "$job_name" || true
```
This is mostly redundant with `kubectl get -o yaml`. This only adds the information from `kubectl events --for=...`. Do that instead.

MikeSpreitzer commented on test/e2e/common/setup-shell.sh:
```diff
@@ -48,3 +48,87 @@ function expect-cmd-output() {
+            echo "--- job.yaml ---" >&2
+            kubectl --context "$context" -n "$namespace" get job "$job_name" -o yaml || true
+            echo "--- pods (wide) ---" >&2
+            kubectl --context "$context" -n "$namespace" get pods -l job-name="$job_name" -o wide || true
+            local pods
+            pods=$(kubectl --context "$context" -n "$namespace" get pods -l job-name="$job_name" -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}' 2>/dev/null || true)
+            for p in $pods; do
+                echo "--- logs: $p (all containers) ---" >&2
+                kubectl --context "$context" -n "$namespace" logs "$p" --all-containers=true --tail=200 || true
+            done
```
This is mostly redundant with the pod and log dumping at the end of the E2E testing workflow. I think it would be better to add anything missing there, rather than have a whole additional bunch of commands here.

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -127,12 +127,12 @@ else
-kubectl wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.install-status-addon}=true' --timeout 400s
-kubectl wait -n its1-system job.batch/install-status-addon --for condition=Complete --timeout 400s
+kubectl --context "$HOSTING_CONTEXT" wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-hub-init}=true' --timeout 400s
+wait-for-job-complete "$HOSTING_CONTEXT" its1-system its-hub-init 400
```
That new command is a lot of complexity for little gain. What it adds should be added to the post-test debug dumping instead.

rishi-jat commented on test/e2e/common/setup-shell.sh:
```diff
@@ -48,3 +48,87 @@ function expect-cmd-output() {
+        else
+            # Check Complete condition
+            local cond
+            cond=$(echo "$job_json" | sed -n 's/.*\("type" *: *"Complete"[^]]*\).*/\1/p' | grep -o '"status" *: *"[^"]*"' | tail -n1 | awk -F'"' '{print $4}')
```
got it.

rishi-jat commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -127,12 +127,12 @@ else
-kubectl wait -n its1-system job.batch/its-hub-init --for condition=Complete --timeout 400s
-kubectl wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.install-status-addon}=true' --timeout 400s
-kubectl wait -n its1-system job.batch/install-status-addon --for condition=Complete --timeout 400s
+kubectl --context "$HOSTING_CONTEXT" wait controlplane.tenancy.kflex.kubestellar.org/its1 --for 'jsonpath={.status.postCreateHooks.its-hub-init}=true' --timeout 400s
```
Got it.

rishi-jat commented:
Hi @MikeSpreitzer ,
I’ve removed the custom job wait logic and now use only simple kubectl wait commands as you suggested. The script is minimal, context usage is consistent, and diagnostics are handled in the workflow. This should fully resolve the E2E flakiness (#3098, #2201). Thanks for your feedback!
